Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp

```diff
@@ -154,31 +161,108 @@ namespace ROS2
         {
             point = inverseLidarTM.TransformPoint(point);
         }
-
+        typedef float PackedVector3[3];
         auto* ros2Frame = Utils::GetGameOrEditorComponent<ROS2FrameComponent>(GetEntity());
         auto message = sensor_msgs::msg::PointCloud2();
+        const auto pointCount = lastScanResults.m_points.size();
         message.header.frame_id = ros2Frame->GetFrameID().data();
         message.header.stamp = ROS2Interface::Get()->GetROSTimestamp();
         message.height = 1;
-        message.width = lastScanResults.m_points.size();
-        message.point_step = sizeof(AZ::Vector3);
-        message.row_step = message.width * message.point_step;
+        message.width = pointCount;
+        message.point_step = sizeof(PackedVector3);
 
-        AZStd::array<const char*, 3> point_field_names = { "x", "y", "z" };
-        for (int i = 0; i < point_field_names.size(); i++)
-        {
+        AZStd::array<const char *, 3> pointFieldNames = {"x", "y", "z"};
+
+        for (int i = 0; i < pointFieldNames.size(); i++) {
             sensor_msgs::msg::PointField pf;
-            pf.name = point_field_names[i];
+            pf.name = pointFieldNames[i];
             pf.offset = i * 4;
             pf.datatype = sensor_msgs::msg::PointField::FLOAT32;
             pf.count = 1;
             message.fields.push_back(pf);
         }
 
-        size_t sizeInBytes = lastScanResults.m_points.size() * sizeof(AZ::Vector3);
+        if (lastScanResults.m_ids.has_value()) {
+            sensor_msgs::msg::PointField pfId;
+            pfId.name = "entity_id";
+            pfId.offset = message.point_step;
+            pfId.datatype = sensor_msgs::msg::PointField::INT32;
+            pfId.count = 1;
+            message.fields.push_back(pfId);
+            constexpr auto fieldLength = sizeof(int32_t);
+            message.point_step += fieldLength;
+        }
+        AZStd::array<AZ::Color, 256> colorLookupTable;
+        if (lastScanResults.m_classes.has_value()) {
+            colorLookupTable = m_lidarCore.m_lidarConfiguration.
+                    GenerateSegmentationColorsLookupTable();
+
+            AZStd::array<const char *, 3> colorFieldNames = {"r", "g", "b"};
+
+            for (int i = 0; i < colorFieldNames.size(); i++) {
+                sensor_msgs::msg::PointField pf;
+                pf.name = colorFieldNames[i];
+                pf.offset = message.point_step;
+                pf.datatype = sensor_msgs::msg::PointField::FLOAT32;
+                pf.count = 1;
+                message.fields.push_back(pf);
+                message.point_step += sizeof(float);
+            }
+
+            sensor_msgs::msg::PointField pfClassId;
```

Comment:
This might be a preference, but I would rather have class if before its color, and right after entity id.

Reply:
This is to have floats of color to be aligned, (like floatx6+uint *1) , but I can change it

Reply:
Yes, please.

Reply:
Ok, it may be left as it is.